File: Backend/services/distance.js
/**
 * Calculate distance between two coordinates using Haversine formula
 * @param {number} lat1 - Latitude of first point
 * @param {number} lon1 - Longitude of first point
 * @param {number} lat2 - Latitude of second point
 * @param {number} lon2 - Longitude of second point
 * @returns {number} Distance in kilometers
 */
function calculateDistance(lat1, lon1, lat2, lon2) {
  const R = 6371; // Radius of the Earth in kilometers
  const dLat = toRad(lat2 - lat1);
  const dLon = toRad(lon2 - lon1);
  
  const a = 
    Math.sin(dLat / 2) * Math.sin(dLat / 2) +
    Math.cos(toRad(lat1)) * Math.cos(toRad(lat2)) *
    Math.sin(dLon / 2) * Math.sin(dLon / 2);
  
  const c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
  const distance = R * c;
  
  return Math.round(distance * 100) / 100; // Round to 2 decimal places
}

/**
 * Convert degrees to radians
 */
function toRad(degrees) {
  return degrees * (Math.PI / 180);
}

/**
 * Find nearby services within a radius
 * @param {Array} services - Array of services with latitude and longitude
 * @param {number} userLat - User's latitude
 * @param {number} userLng - User's longitude
 * @param {number} radiusKm - Radius in kilometers (default: 10km)
 * @returns {Array} Sorted array of services with distance property
 */
function findNearby(services, userLat, userLng, radiusKm = 10) {
  if (!userLat || !userLng) {
    return [];
  }

  const nearby = services
    .filter(service => service.latitude && service.longitude)
    .map(service => {
      const distance = calculateDistance(
        userLat,
        userLng,
        service.latitude,
        service.longitude
      );
      return {
        ...service,
        distance
      };
    })
    .filter(service => service.distance <= radiusKm)
    .sort((a, b) => a.distance - b.distance);

  return nearby;
}

module.exports = {
  calculateDistance,
  findNearby,
  toRad
};

File: Backend/services/distance.py
import math

R = 6371  # Earth radius in kilometers


def to_rad(degrees):
    return degrees * (math.pi / 180)


def calculate_distance(lat1, lon1, lat2, lon2):
    """Return distance in kilometers between two coordinates (Haversine)."""
    dlat = to_rad(lat2 - lat1)
    dlon = to_rad(lon2 - lon1)
    a = (math.sin(dlat / 2) ** 2 + math.cos(to_rad(lat1)) * math.cos(to_rad(lat2)) * math.sin(dlon / 2) ** 2)
    c = 2 * math.atan2(math.sqrt(a), math.sqrt(1 - a))
    distance = R * c
    return round(distance, 2)


def find_nearby(services, user_lat, user_lng, radius_km=10):
    try:
        user_lat = float(user_lat)
        user_lng = float(user_lng)
    except Exception:
        return []

    nearby = []
    for s in services:
        lat = s.get('latitude')
        lng = s.get('longitude')
        if lat is None or lng is None:
            continue
        try:
            dist = calculate_distance(user_lat, user_lng, float(lat), float(lng))
        except Exception:
            continue
        new_s = dict(s)
        new_s['distance'] = dist
        if dist <= float(radius_km):
            nearby.append(new_s)

    nearby.sort(key=lambda x: x.get('distance', float('inf')))
    return nearby
